src/Features/PlayerTrace.cpp

```diff
@@ -79,48 +96,40 @@ void PlayerTrace::DrawInWorld(float time) const
-                // Display speed delta between first and last groundframe
-                speed_a = (trace[first_groundframe_tick - idx_offset] - trace[first_groundframe_tick])/idx_offset;
-                speed_b = (trace[last_groundframe_tick - idx_offset] - trace[last_groundframe_tick])/idx_offset;
-                speed_delta = sqrt(speed_b.x*speed_b.x + speed_b.y*speed_b.y) - sqrt(speed_a.x*speed_a.x + speed_a.y*speed_a.y);
-
-                engine->PointToScreen(trace[first_groundframe_tick + (last_groundframe_tick - first_groundframe_tick)/2] + hud_offset, screen_pos);
-                ctx->DrawElementOnScreen(hud_id++, screen_pos.x, screen_pos.y, "%10.2f", speed_delta * 60);
+    switch (groundframe_swaps.size()) {
+        case 0:
+            DrawSpeedDelta(ctx, &hud_id, 0, trace.size()-1);
+            break;
+        case 1:
+            DrawSpeedDelta(ctx, &hud_id, 0, groundframe_swaps[0]);
+            DrawSpeedDelta(ctx, &hud_id, groundframe_swaps[0], trace.size()-1);
+            break;
+        default:
+            DrawSpeedDelta(ctx, &hud_id, 0, groundframe_swaps[0]);
+            DrawSpeedDelta(ctx, &hud_id, groundframe_swaps[groundframe_swaps.size()-1], trace.size()-1);
+            for (size_t i = 1; i < groundframe_swaps.size(); i++) {
+                DrawSpeedDelta(ctx, &hud_id, groundframe_swaps[i-1], groundframe_swaps[i]);
             }
-            // Display speed delta between last groundframe and current tick
-            // aka speed delta for the current hop
-            speed_a = (trace[last_groundframe_tick - idx_offset] - trace[last_groundframe_tick])/idx_offset;
-            speed_b = (trace[groundframes[i] - idx_offset] - trace[groundframes[i]])/idx_offset;
-            speed_delta = sqrt(speed_b.x*speed_b.x + speed_b.y*speed_b.y) - sqrt(speed_a.x*speed_a.x + speed_a.y*speed_a.y);
-
-            engine->PointToScreen(trace[last_groundframe_tick + (groundframes[i] - last_groundframe_tick)/2] + hud_offset, screen_pos);
-            ctx->DrawElementOnScreen(hud_id++, screen_pos.x, screen_pos.y, "%10.2f", speed_delta * 60);
-
-            first_groundframe_tick = groundframes[i];
-            last_groundframe_tick = groundframes[i];
-        }
+            break;
     }
 }
+void PlayerTrace::DrawSpeedDelta(HudContext *ctx, int *hud_id, size_t begin_tick, size_t end_tick) const {
+    const Vector hud_offset = {0.0, 0.0, 10.0};
+    const size_t idx_offset = 2; // To account for alt ticks
+    Vector screen_pos;
+
+    if (begin_tick == 0) begin_tick += idx_offset;
+    Vector speed_a = (trace[begin_tick - idx_offset] - trace[begin_tick])/idx_offset;
+    Vector speed_b = (trace[end_tick - idx_offset] - trace[end_tick])/idx_offset;
+    float speed_delta = speed_b.Length2D() - speed_a.Length2D();
+
+    engine->PointToScreen(trace[begin_tick + (end_tick - begin_tick)/2] + hud_offset, screen_pos);
+    ctx->DrawElementOnScreen((*hud_id)++, screen_pos.x, screen_pos.y, "%10.2f", speed_delta * 60);
+}
```

Comment:
there are a lot of words to read in this function and i am very sleepy so i am going to assume all your maths is perfect